crates/matrix-sdk/src/widget/machine/pending.rs

```diff
@@ -0,0 +1,173 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A wrapper around a hash map that tracks pending requests and makes sure
+//! that expired requests are removed.
+
+use std::time::{Duration, Instant};
+
+use indexmap::{map::Entry, IndexMap};
+use uuid::Uuid;
+
+/// Configuration of limits for the outgoing request handling.
+#[derive(Clone, Debug)]
+pub(crate) struct RequestLimits {
+    /// Maximum amount of unanswered (pending) requests that the client widget
+    /// API is going to process before starting to drop them. This ensures
+    /// that a buggy widget cannot force the client machine to consume memory
+    /// indefinitely.
+    pub(crate) max_pending_requests: usize,
+    /// For how long can the unanswered (pending) request stored in a map before
+    /// it is dropped. This ensures that requests that are not answered within
+    /// a ceratin amount of time, are dropped/cleaned up (considered as failed).
+    pub(crate) response_timeout: Duration,
+}
+
+/// A wrapper around a hash map that ensures that the request limits
+/// are taken into account (expired requests get cleaned up and that
+/// the hashmap remains limited to a certain amount of pending requests).
+pub(super) struct PendingRequests<T> {
+    requests: IndexMap<Uuid, Expirable<T>>,
+    limits: RequestLimits,
+}
+
+impl<T> PendingRequests<T> {
+    pub(super) fn new(limits: RequestLimits) -> Self {
+        Self { requests: IndexMap::with_capacity(limits.max_pending_requests), limits }
+    }
+
+    /// Inserts a new request into the map.
+    /// Returns `None` if the maximum allowed capacity is reached.
+    pub(super) fn insert(&mut self, key: Uuid, value: T) -> Option<&mut T> {
+        if self.requests.len() >= self.limits.max_pending_requests {
+            return None;
+        }
+
+        let Entry::Vacant(entry) = self.requests.entry(key) else {
+            panic!("uuid collision");
+        };
+
+        let expirable = Expirable::new(value, Instant::now() + self.limits.response_timeout);
+        let inserted = entry.insert(expirable);
+        Some(&mut inserted.value)
+    }
+
+    /// Extracts a request from the map based on its identifier.
+    /// Returns `None` if the value is not present or expired.
+    pub(super) fn extract(&mut self, key: &Uuid) -> Result<T, &'static str> {
+        let value = self.requests.remove(key).ok_or("Received response for an unknown request")?;
+        value.value().ok_or("Dropping response for an expired request")
+    }
+
+    /// Removes all expired requests from the map.
+    pub(super) fn remove_expired(&mut self) {
+        self.requests.retain(|_, v| !v.expired());
```

Comment:
I think we should log when requests are dropped because they expired.

Reply:
Updated.